File: src/main/java/dark/leech/text/plugin/RepositoryManager.java
package dark.leech.text.plugin;

import java.util.ArrayList;
import java.util.List;

import org.apache.commons.collections4.CollectionUtils;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import dark.leech.text.action.Log;
import dark.leech.text.enities.RepositoryEntity;
import dark.leech.text.util.AppUtils;
import dark.leech.text.util.FileUtils;

public class RepositoryManager {
    private static final Gson gson = new Gson();

    private static RepositoryManager manager;

    private static List<RepositoryEntity> repositoryList;

    private RepositoryManager() {

        List<RepositoryEntity> repo = new ArrayList<>();
        try {
            var json = FileUtils.file2string(AppUtils.curDir + "/tools/repository.json");
            var type = TypeToken.getParameterized(List.class, RepositoryEntity.class).getType();
            repo = gson.fromJson(json, type);
        } catch (Exception e) {
            Log.add(e);
        }

        if (CollectionUtils.isNotEmpty(repo)) repositoryList = repo;
    }

    public static RepositoryManager getManager() {
        if (manager == null) manager = new RepositoryManager();
        return manager;
    }

    public boolean hasRepoSetting() {
        return FileUtils.file2string(AppUtils.curDir + "/tools/repository.json") != null;
    }

    public List<RepositoryEntity> repositoryList() {
        return repositoryList;
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/plugin/RepositoryManager.kt
package dev.haipham22.leechtext.plugin

import dev.haipham22.leechtext.entities.RepositoryEntity
import dev.haipham22.leechtext.log.EngineLogger
import dev.haipham22.leechtext.util.EnginePaths
import dev.haipham22.leechtext.util.LeechJson
import dev.haipham22.leechtext.util.readTextOrNull
import dev.haipham22.leechtext.util.writeTo
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

/**
 * Quản lý repo plugin đã lưu local (port từ plugin/RepositoryManager.java + RepositoryUI).
 * DI class (Koin singleton), đọc tools/repository.json (format: Gson List&lt;RepositoryEntity&gt; như bản gốc).
 * P3 thêm add/remove + persist (MUST #17).
 */
@OptIn(ExperimentalUuidApi::class)
class RepositoryManager(
    private val log: EngineLogger,
) {
    private val repositoryList: MutableList<RepositoryEntity> = loadRepositories()

    private fun loadRepositories(): MutableList<RepositoryEntity> {
        // Chưa từng lưu repo là trạng thái hợp lệ — trả rỗng, không log stack
        if (!okio.FileSystem.SYSTEM.exists(repositoryFile())) return ArrayList()
        return try {
            val json = repositoryFile().readTextOrNull() ?: return ArrayList()
            LeechJson.decodeFromString<List<RepositoryEntity>>(json).toMutableList()
        } catch (e: Exception) {
            log.add(e)
            ArrayList()
        }
    }

    private fun repositoryFile(): okio.Path = EnginePaths.dataDir / "tools" / "repository.json"

    /** Có thiết lập repo chưa (file repository.json tồn tại + đọc được). */
    fun hasRepoSetting(): Boolean = try {
        repositoryFile().readTextOrNull()?.isNotEmpty() == true
    } catch (e: Exception) {
        false
    }

    fun repositoryList(): List<RepositoryEntity> = repositoryList

    /** Thêm repo theo link (enabled). Trả false nếu link rác hoặc đã có. */
    fun add(link: String): Boolean {
        val trimmed = link.trim()
        if (!trimmed.startsWith("http")) return false
        if (repositoryList.any { it.link == trimmed }) return false
        val entity =
            RepositoryEntity(
                uuid = Uuid.random().toString(),
                link = trimmed,
                isEnabled = true,
            )
        repositoryList.add(entity)
        persist()
        return true
    }

    fun remove(entity: RepositoryEntity) {
        repositoryList.removeAll { it.link == entity.link }
        persist()
    }

    fun setEnabled(
        entity: RepositoryEntity,
        enabled: Boolean,
    ) {
        repositoryList.firstOrNull { it.link == entity.link }?.isEnabled = enabled
        persist()
    }

    /** Ghi repository.json — format Gson list như bản gốc. */
    private fun persist() {
        try {
            LeechJson.encodeToString(repositoryList).writeTo(repositoryFile().toString(), log = log)
        } catch (e: Exception) {
            log.add(e)
        }
    }
}
